src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -359,6 +380,158 @@ def to_Spec {f : A} (m : ℕ) (f_deg : f ∈ 𝒜 m) :
+  simp_rw [mk_mul, ← pow_add, map_add, add_pow, mk_sum, mul_comm, ← nsmul_eq_mul, ← smul_mk],
+  let g : ℕ → A⁰_ f_deg := λ j, (m + m).choose j • if h2 : m + m < j then 0 else if h1 : j ≤ m
+    then ⟨mk (proj 𝒜 i a ^ j * proj 𝒜 i b ^ (m - j)) ⟨_, i, rfl⟩, i, ⟨_, _⟩, rfl⟩ *
+      ⟨mk (proj 𝒜 i b ^ m) ⟨_, i, rfl⟩, i, ⟨_, by mem_tac⟩, rfl⟩
+    else ⟨mk (proj 𝒜 i a ^ m) ⟨_, i, rfl⟩, i, ⟨_, by mem_tac⟩, rfl⟩ *
+      ⟨mk (proj 𝒜 i a ^ (j - m) * proj 𝒜 i b ^ (m + m - j)) ⟨_, i, rfl⟩, i, ⟨_, _⟩, rfl⟩,
+  rotate,
+  { rw (_ : m * i = _), mem_tac, rw [← add_smul, nat.add_sub_of_le h1], refl },
+  { rw (_ : m * i = _), mem_tac, rw ← add_smul, congr,
+    zify [le_of_not_lt h2, le_of_not_le h1], abel },
+  convert_to ∑ i in range (m + m + 1), g i ∈ q.1, swap,
+  { refine q.1.sum_mem (λ j hj, nsmul_mem _ _), split_ifs,
+    exacts [q.1.zero_mem, q.1.mul_mem_left _ (hb i), q.1.mul_mem_right _ (ha i)] },
+  apply subtype.ext,
+  rw [degree_zero_part.coe_sum, subtype.coe_mk],
+  apply finset.sum_congr rfl (λ j hj, _),
+  congr' 1, split_ifs with h2 h1,
+  { exact ((finset.mem_range.1 hj).not_le h2).elim },
+  all_goals { simp only [subtype.val_eq_coe, degree_zero_part.coe_mul, subtype.coe_mk, mk_mul] },
+  { rw [mul_assoc, ← pow_add, add_comm (m - j), nat.add_sub_assoc h1] },
+  { rw [← mul_assoc, ← pow_add, nat.add_sub_of_le (le_of_not_le h1)] },
+end
+
+lemma carrier.smul_mem (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) (c x : A) (hx : x ∈ carrier q) :
+  c • x ∈ carrier q :=
+begin
+  revert c,
+  refine direct_sum.decomposition.induction_on 𝒜 _ _ _,
+  { rw zero_smul, exact carrier.zero_mem hm _ },
+  { rintros n ⟨a, ha⟩ i,
+    simp_rw [subtype.coe_mk, proj_apply, smul_eq_mul, coe_decompose_mul_of_left_mem _ ha],
+    split_ifs,
+    { convert_to (⟨mk _ ⟨_, n, rfl⟩, n, ⟨_, pow_mem_graded m ha⟩, rfl⟩ : A⁰_ f_deg) *
+        ⟨mk _ ⟨_, i - n, rfl⟩, _, ⟨proj 𝒜 (i - n) x ^ m, by mem_tac⟩, rfl⟩ ∈ q.1,
+      { erw [subtype.ext_iff, subring.coe_mul, mk_mul, subtype.coe_mk, mul_pow],
+        congr, erw [← pow_add, nat.add_sub_of_le h] },
+      { exact ideal.mul_mem_left _ _ (hx _) } },
+    { simp_rw [zero_pow hm, mk_zero], exact q.1.zero_mem } },
+  { simp_rw add_smul, exact λ _ _, carrier.add_mem q },
+end
+
+/--
+For a prime ideal `q` in `A⁰_f`, the set `{a | aᵢᵐ/fⁱ ∈ q}` as an ideal.
+-/
+def carrier.as_ideal (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) :
+  ideal A :=
+{ carrier := carrier q,
+  zero_mem' := carrier.zero_mem hm q,
+  add_mem' := λ a b, carrier.add_mem q,
+  smul_mem' := carrier.smul_mem hm q }
+
+lemma carrier.as_ideal.ne_top (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) :
+  (carrier.as_ideal hm q) ≠ ⊤ :=
+λ rid, q.2.ne_top $ (ideal.eq_top_iff_one _).mpr
+begin
+  convert (ideal.eq_top_iff_one _).mp rid 0,
+  simpa only [subtype.ext_iff, degree_zero_part.coe_one, subtype.coe_mk,
+    proj_apply, decompose_of_mem_same 𝒜 one_mem, one_pow, pow_zero, ← mk_self 1],
+end
+
+lemma carrier.as_ideal.homogeneous  (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) :
```

Comment:
Actually if we switch the order of `zero_mem` and `add_mem`, then we can use `variables (hm : 0 < m) (q : Spec.T (A⁰_ f_deg))` (and `include hm`) for everything below and including `zero_mem`. What do you think?
```suggestion
lemma carrier.as_ideal.homogeneous (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) :
```

Reply:
Good idea! These changes has been applied